cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+for either of these use cases, it will demonstrate the benefits of having an
+xDS adapter when developing a solution for either of these use cases.
+
+### Use Cases
+
+#### Topology Aware Routing
+
+As the author of both EndpointSlices and the existing Topology Aware Routing
+approach in Kubernetes, I feel like I’m well equipped to discuss the limitations
+of both. What most people want from Topology Aware Routing is essentially to
+fill the local zone to capacity and then spillover to the next closest set of
+endpoints with capacity. Unfortunately that is very difficult to achieve with
+the existing Kubernetes tooling. There are a few key problems here:
+
+1. Kubernetes has no feedback loop to its dataplane and therefore no reliable
+   way to understand if endpoints have reached capacity.
+2. Even if there were a feedback loop, spillover to other zones benefits from
+   some form of centralized orchestration to avoid a thundering herd problem.
+3. All changes to endpoint routing currently have to go through Services or
+   EndpointSlices, and by extension, the Kubernetes API Server. It would be very
+   expensive to implement incremental weighting/spillover adjustments with that
+   overhead. For example, adjusting the weight of any individual endpoint would
+   require a write to the EndpointSlice API which would then need to be
+   distributed to all consumers of that API. That can be problematic from a
+   scalability perspective when there are frequent updates paired with a large
+   number of endpoints and/or nodes.
+
+I believe that xDS is uniquely positioned to address these limitations.
+[LRS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/load_stats/v3/lrs.proto)
+(Load Reporting Service) provides a straightforward way to provide load reports
+to a centralized control plane. Similarly, [Delta
+xDS](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/xds_api#delta-endpoints)
+enables xDS to distribute incremental updates that can be smaller in scope than
+comparable Kubernetes APIs. All of this could be combined to avoid adding
+additional load on the Kubernetes API Server.
+
+#### Endpoints From Other Sources
+
+In some cases, it’s useful to be able to route to endpoints from other sources,
+such as other Kubernetes clusters or outside of Kubernetes altogether. This can
+be particularly relevant when implementing Multi-Cluster Service routing.
+Implementations today often rely on one of the following approaches:
+
+1. Mirror endpoints from other clusters into Kubernetes by creating custom
+   Services and EndpointSlices.
+2. Introduce Gateways on the edges of clusters that can be used to route
+   requests from other clusters.
+
+With xDS support, it could be possible to have a centralized xDS control plane
+that was familiar with endpoints in multiple sources, providing a potentially
+more efficient alternative to either of the approaches commonly used today.
```

Comment:
What is it about this centralized xDS control plane that is more efficient?

Probably worth also contrasting with kvstoremesh (see CiliumCon or https://arthurchiao.art/blog/trip-first-step-towards-cloud-native-security/#222-kvstoremesh)

Reply:
https://github.com/cilium/cilium/issues/30283 describes the benefits of centralization in a different context. Just wanted to share this for reviewers not aware of the adjacent proposal.